crates/sui-core/src/checkpoints/causal_order.rs

```diff
@@ -0,0 +1,421 @@
+                // We only record the dependencies not in a checkpoint, as the ones
+                // in a checkpoint are already satisfied presumably.
+                if tx_not_in_checkpoint.contains(dep) {
+                    forward_index.entry(dep).or_default().push(d);
+                }
+            }
+        }
+
+        // Define the master sequence, to contain the initial transactions
+        // by transaction digest order.
+        let mut master_sequence: Vec<&TransactionDigest> =
+            initial_transactions.iter().cloned().collect();
+        // A set mirroring the contents of the mater sequence for quick lookup
+        let mut master_set = initial_transactions.clone();
+        // The transactions that just became executed.
+        let mut candidates = initial_transactions;
+
+        // Trace forward the executed transactions, starting from the initial set, and adding more
+        // trasnactions as all their dependencies become executed. The candidates represent executed
+        // transactions that need to have subsequent transactions depending on them examined to
+        // determine if all their dependencies are executed. If so they are sequenced, and also added
+        // to the candidate set to be examiner once.
+        while !candidates.is_empty() {
+            // we continue while we can make progress
+
+            // Take a transaction
+            let next_transaction = *candidates.iter().next().unwrap();
+            candidates.remove(next_transaction);
+
+            // If nothing depends on this tx move on to the next
+            let forward_deps = forward_index.get(next_transaction);
+            if forward_deps.is_none() {
+                continue;
+            }
+
+            // Check all transactions that depend on it, to see if all
+            // dependencies are  satisfied.
+            for dep in forward_deps.unwrap() {
+                // If the forward dependency is not included in the current checkpoint, ignore.
+                if !effect_map.contains_key(*dep) {
+                    continue;
+                }
+
+                // We have already included this into the candidate set once.
+                if master_set.contains(dep) {
+                    continue;
+                }
+
+                if effect_map
+                    .get(*dep)
+                    .unwrap()
+                    .dependencies
+                    .iter()
+                    // All dependencies sequenced in the master seq or in previous checkpoint
+                    .filter(|item| tx_not_in_checkpoint.contains(*item))
+                    .all(|item| master_set.contains(item))
+                {
+                    // It seems like all dependencies are satisfied for dep, so sequence it.
+                    master_sequence.push(dep);
+                    master_set.insert(dep);
+                    candidates.insert(dep);
+                }
+            }
+        }
```

Comment:
I believe this topo sort works, but I think a more traditional approach may be simpler and faster:

```
let mut to_visit: BTreeSet<&TxDigest> = ...; // populate with all txes.
let mut in_degree: HashMap<&TxDigest, usize> = HashMap::new();

// can probably compute in_degree when we build forward_index
digests.for_each(|digest| {
    forward_index[digest].iter()
       .for_each(|child| {
            let mut entry = in_degree.entry(child).or_insert(0);
            to_visit.remove(child);
            entry += 1;
    })
});

// in_degree now has the number of parents of every tx with parents in this ckpt
// to_visit contains only txes with no parents.

let toposorted = Vec::new();

// could avoid converting to vector with: https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#method.pop_first
let to_visit: Vec<_> = to_visit.iter().collect();

while let Some(digest) = to_visit.pop() {
   toposorted.push(digest);
   forward_index.get(digest).for_each(|child| {
       if let Entry::Occupied(entry) = in_degree.entry(child) {
          *entry -= 1;
          if *entry == 0 {
             toposorted.push(child);
             entry.remove();
             to_visit.push(child);
          }
       }
    }
}
```

Obviously didn't test that so probably has awful bugs but that's the idea.

Reply:
Just implemented the above (with small bugs removed) and it is indeed much more clear.